### PR TITLE
[FLINK-20580][core] Don't support nullable value for SerializedValue

### PR DESCRIPTION
**Pull request by kezhuw**
## What is the purpose of the change
Dropping nullable-value support from `SerializedValue` to simplify its usages. Callers should resort to nullabe/optional variable of `SerializedValue` if them want such behavior but just can't treat `SerializedValue` as another optional container anymore.

## Brief change log

  - Migrate usage of `SerializedValue` in rpc module to its own wire value class. This make this module self-contained and resistant to rpc usages of `SerializedValue`.
  - Dropping nullable-value support from `SerializedValue`.

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.
* `SerializedValueTest.testSimpleValue`.

This change added tests and can be verified as follows:
- `SerializedValueTest.testNullValue`, `SerializedValueTest.testFromNullBytes`, `SerializedValueTest.testFromEmptyBytes`
- `AkkaRpcSerializedValueTest`
- `AkkaRpcActorTest.canRespondWithSerializedValueLocally`, `RemoteAkkaRpcActorTest.canRespondWithSerializedValueRemotely`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

## WARNING
**This change will breaking tm/jm rolling update if we ever support it.** I got nothing of such supporting from docs and assume no such guarantee.

### CODE REVIEW

**Comment by kezhuw** [comment]
Azure [FAILURE](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=13308&view=logs&j=c5f0071e-1851-543e-9a45-9ac140befc32&t=1fb1a56f-e8b5-5a82-00a0-a2db7757b4f5&l=12861) is unrelated to changes and reported to existing jira https://issues.apache.org/jira/browse/FLINK-21214.

**Comment by kezhuw** [comment]
@flinkbot run azure

**Comment by kezhuw** [comment]
Hi @rmetzger, `@flinkbot run azure` does not function anymore ?

**Comment by kezhuw** [comment]
@flinkbot attention @tillrohrmann

**Comment by rmetzger** [comment]
The most reliable way of triggering another Azure build is by pushing to the branch again. It seemed to have worked this time.
I'll ask Chesnay to look into the reason why the command didn't work.

File: flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java [hunk 1]
```diff
@@ -47,26 +50,47 @@ public void testSimpleValue() {
...
+                    SerializedValue.fromBytes(Arrays.copyOf(bytes, bytes.length));
+            assertEquals(v, saved);
+            assertArrayEquals(v.getByteArray(), saved.getByteArray());
         } catch (Exception e) {
             e.printStackTrace();
             fail(e.getMessage());
         }
     }
 
     @Test
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java:67]: Might be a bit shorter:

```suggestion
    @Test(expected = NullPointerException.class)
```

File: flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java [hunk 2]
```diff
@@ -47,26 +50,47 @@ public void testSimpleValue() {
...
-            SerializedValue<Object> copy = CommonTestUtils.createCopySerializable(v);
+            new SerializedValue<>(null);
+            fail("expect NullPointerException");
+        } catch (NullPointerException e) {
+            // ignore
+        }
+    }
 
-            assertNull(copy.deserializeValue(getClass().getClassLoader()));
+    @Test
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java:77]: Same here with `@Test(expected = ...)`.

File: flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java [hunk 3]
```diff
@@ -47,26 +50,47 @@ public void testSimpleValue() {
...
+        }
+    }
 
-            assertEquals(v, copy);
-            assertEquals(v.hashCode(), copy.hashCode());
-            assertEquals(v.toString(), copy.toString());
-        } catch (Exception e) {
-            e.printStackTrace();
-            fail(e.getMessage());
+    @Test
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/util/SerializedValueTest.java:87]: Same here with `@Test(exepcted = ...)`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSerializedValueTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+
+public class AkkaRpcSerializedValueTest {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSerializedValueTest.java:39]:
```suggestion
public class AkkaRpcSerializedValueTest extends TestLogger {
```

This makes sure that the test names are printed. This is quite helpful for CI where multiple tests are executed one after another and the names help to split the log messages from the different test cases.

**Review by tillrohrmann — APPROVED** [review]
Very good work @kezhuw! The changes look good to me. I will address my own comments while merging this PR.

**Comment by kezhuw** [comment]
@tillrohrmann Thanks for reviewing, the comments make sense to me. Then I will give up rebasing to you.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSerializedValue.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Arrays;
+
+/** A self-contained serialized value to decouple from user values and transfer on wire. */
+class AkkaRpcSerializedValue implements Serializable {
+    @Nullable private final byte[] serializedData;
```
Comment (kezhuw) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSerializedValue.java:32]: @tillrohrmann Missed `serialVersionUID`.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSerializedValue.java:32]: Good point. Will add it as well.

**Pull request closed without merging**